### PR TITLE
Gift exchange added!

### PR DESCRIPTION
**Pull request by zarutian**
So people can pass various things between themselfs for purposes of making complex examples.

ps. I hope difference of coding style is not too clashing.

### CODE REVIEW

**Comment by dckc** [comment]
I don't see any reason to use a completely different style from the rest of the code:

> Monte’s blocks are usually indentation-delimited: ...
> Monte code should always uses four spaces for each indentation level.
 -- http://monte.readthedocs.io/en/latest/symbols.html#brackets-indentation-and-blocks

File: airbrus.mt [hunk 1]
```diff
@@ -68,6 +68,73 @@ def UTF8JSON :DeepFrozen := composeCodec(UTF8, JSON)
...
+    def wants := wanters.fetch(recipiant)
+    wants.set(giftName, promres)
+    return promise
+  }
+  
+  def giftExchange_interface_maker(nickname :Str) :Any {
+    object GEI {
+      """Your interface to the giftExchange"""
+      to give(recipiant :Str, giftName :Str, gift :Any) :Void {
+        """signiture: give(recipiant :Str, giftName :Str, gift :Any) :Void"""
```
Comment (dckc) [inline airbrus.mt]: typo: signature

File: airbrus.mt [hunk 2]
```diff
@@ -68,6 +68,73 @@ def UTF8JSON :DeepFrozen := composeCodec(UTF8, JSON)
...
+  def givers  := [].asMap().diverge()
+  def wanters := [].asMap().diverge()
+  object duck {}
+  def pass_duck () :Any {
+    return duck
+  }
+  
+  def give(giver :Str, recipiant :Str, giftName :Str, gift :Any) :Void {
+    def giftBox := gift  # Allways wrap your presents! (For when gifts are promises)
+    if (wanters.fetch(recipiant, pass_duck) != duck) {
```
Comment (dckc) [inline airbrus.mt]: typo: recipient

File: airbrus.mt [hunk 3]
```diff
@@ -68,6 +68,62 @@ def UTF8JSON :DeepFrozen := composeCodec(UTF8, JSON)
...
     return makePumpTube(makeSplitPump(b`$\n`))
 
+def make_giftExchange() :Any as DeepFrozen:
+    def givers  := [].asMap().diverge()
+    def wanters := [].asMap().diverge()
+    object duck {}
+    def pass_duck () :Any :
+        return duck
+        
+    def give(giver :Str, recipiant :Str, giftName :Str, gift :Any) :Void :
```
Comment (dckc) [inline airbrus.mt]: typo: recipient

File: airbrus.mt [hunk 4]
```diff
@@ -68,6 +68,62 @@ def UTF8JSON :DeepFrozen := composeCodec(UTF8, JSON)
...
+              return
+        if (givers.fetch(giver, pass_duck) == duck):
+            givers[giver] := [].asMap().diverge()
+        def gifts := givers[giver]
+        gifts[giftName] := &giftBox
+        return
+  
+    def accept(recipiant :Str, giver :Str, giftName :Str) :Any :
+        def promres := Ref.make_promise()
+        var [promise, resolver] := promres
```
Comment (dckc) [inline airbrus.mt]: consider a more concise:
```
    def [var promise, var resolver] := Ref.make_promise()
```

Do they need to be var? Let's see...

File: airbrus.mt [hunk 5]
```diff
@@ -68,6 +68,62 @@ def UTF8JSON :DeepFrozen := composeCodec(UTF8, JSON)
...
+        wants.set(giftName, promres)
+        return promise
+  
+    def giftExchange_interface_maker(nickname :Str) :Any :
+        object GEI:
+            """Your interface to the giftExchange"""
+            to give(recipiant :Str, giftName :Str, gift :Any) :Void :
+                """signiture: give(recipiant :Str, giftName :Str, gift :Any) :Void"""
+                give(nickname, recipiant, giftName, gift)
+                return
```
Comment (dckc) [inline airbrus.mt]: What is this `return` for?

File: airbrus.mt [hunk 6]
```diff
@@ -68,6 +68,62 @@ def UTF8JSON :DeepFrozen := composeCodec(UTF8, JSON)
...
+            wanters.set(acceptor, [].asMap().diverge())
+        def wants := wanters.fetch(acceptor)
+        wants.set(giftName, promres)
+        return promise
+  
+    def giftExchange_interface_maker(nickname :Str) :Any :
+        object GEI:
+            """Your interface to the giftExchange"""
+            to give(acceptor :Str, giftName :Str, gift :Any) :Void :
+                """signiture: give(acceptor :Str, giftName :Str, gift :Any) :Void"""
```
Comment (dckc) [inline airbrus.mt:79]: typo: signature

File: airbrus.mt [hunk 7]
```diff
@@ -188,7 +244,7 @@ def main(argv, => Timer,
                 def brusHelp := makeAirbrusHelp(fn s {client.say(channel, s)})
                 def instanceEnv := ["&&help" => &&brusHelp]
                 def userEnv := userEnvironments.fetch(user.getNick(),
-                                                      fn {baseEnv | instanceEnv})
+                                                      fn { return baseEnv | instanceEnv | ["giftExchange" => make_giftExchange_interface(user.getNick())] })
```
Comment (dckc) [inline airbrus.mt]: why add `return` in the `fn`? it's implicit.

Comment (zarutian) [reply airbrus.mt]: been bitten too many times in various other programming languages.

It is also a holdover from when I was playing with E. (Which does not have implict return iirc)

File: airbrus.mt [hunk 8]
```diff
@@ -68,6 +68,55 @@ def UTF8JSON :DeepFrozen := composeCodec(UTF8, JSON)
 def makeLineTube() as DeepFrozen:
     return makePumpTube(makeSplitPump(b`$\n`))
 
+def make_giftExchange() :Any as DeepFrozen:
```
Comment (MostAwesomeDude) [inline airbrus.mt]: Style nit: `makeGiftExchange`

File: airbrus.mt [hunk 9]
```diff
@@ -68,6 +68,55 @@ def UTF8JSON :DeepFrozen := composeCodec(UTF8, JSON)
...
+    def givers  := [].asMap().diverge()
+    def wanters := [].asMap().diverge()
+        
+    def give(giver :Str, acceptor :Str, giftName :Str, gift :Any) :Void :
+        def giftBox := gift  # Allways wrap your presents! (For when gifts are promises)
```
Comment (MostAwesomeDude) [inline airbrus.mt]: This does nothing. If you want to ensure that `gift` is near, guard it as `gift :Near`.

File: airbrus.mt [hunk 10]
```diff
@@ -68,6 +68,55 @@ def UTF8JSON :DeepFrozen := composeCodec(UTF8, JSON)
...
+        if (wanters.contains(acceptor):
```
Comment (MostAwesomeDude) [inline airbrus.mt]: Syntax error; missing closing parenthesis.

File: airbrus.mt [hunk 11]
```diff
@@ -68,6 +68,55 @@ def UTF8JSON :DeepFrozen := composeCodec(UTF8, JSON)
...
+            def wants := wanters[acceptor]
+            if (wants.contains(giftName)):
+              wants.fetch(giftName)[1].resolve(&giftBox)
```
Comment (MostAwesomeDude) [inline airbrus.mt]: This `&giftBox` can just be `&gift`. So can the other one.

File: airbrus.mt [hunk 12]
```diff
@@ -68,6 +68,55 @@ def UTF8JSON :DeepFrozen := composeCodec(UTF8, JSON)
...
+              return
+        if (!givers.contains(giver)):
+            givers[giver] := [].asMap().diverge()
+        def gifts := givers[giver]
+        gifts[giftName] := &giftBox
+        return
+  
+    def accept(acceptor :Str, giver :Str, giftName :Str) :Any :
+        def promres := Ref.make_promise()
```
Comment (MostAwesomeDude) [inline airbrus.mt]: Did you mean `Ref.promise()`?

File: airbrus.mt [hunk 13]
```diff
@@ -68,6 +68,55 @@ def UTF8JSON :DeepFrozen := composeCodec(UTF8, JSON)
...
+        var [promise, resolver] := promres
```
Comment (MostAwesomeDude) [inline airbrus.mt]: Should be folded with the previous line. Also `var` here doesn't actually create any mutable names; did you want `def [var promise, var resolver] := promres`? Also, for style, these names are traditionally `p` and `r`.

File: airbrus.mt [hunk 14]
```diff
@@ -68,6 +68,55 @@ def UTF8JSON :DeepFrozen := composeCodec(UTF8, JSON)
...
+        if (wanters.contains(acceptor):
+            def wants := wanters[acceptor]
+            if (wants.contains(giftName)):
+              wants.fetch(giftName)[1].resolve(&giftBox)
+              return
+        if (!givers.contains(giver)):
+            givers[giver] := [].asMap().diverge()
+        def gifts := givers[giver]
+        gifts[giftName] := &giftBox
+        return
```
Comment (MostAwesomeDude) [inline airbrus.mt]: Not needed.

File: airbrus.mt [hunk 15]
```diff
@@ -68,6 +68,55 @@ def UTF8JSON :DeepFrozen := composeCodec(UTF8, JSON)
...
+  
+    def accept(acceptor :Str, giver :Str, giftName :Str) :Any :
+        def promres := Ref.make_promise()
+        var [promise, resolver] := promres
+        if (givers.contains(giver)):
+            def gifts := givers[giver]
+            if (gifts.contains(giftName)):
+                resolver.resolve(gifts.fetch(giftName)
```
Comment (MostAwesomeDude) [inline airbrus.mt]: Syntax error.

File: airbrus.mt [hunk 16]
```diff
@@ -68,6 +68,55 @@ def UTF8JSON :DeepFrozen := composeCodec(UTF8, JSON)
...
+                return promise
+        if (wanters.contains(acceptor)):
+            def wants := wanters[acceptor]
+            if (wants.contains(giftName)):
+                [promise, resolver] = wants.fetch(giftName)
```
Comment (MostAwesomeDude) [inline airbrus.mt]: This assignment isn't legal.

File: airbrus.mt [hunk 17]
```diff
@@ -68,6 +68,55 @@ def UTF8JSON :DeepFrozen := composeCodec(UTF8, JSON)
...
+                return promise
+        if (!wanters.contains(acceptor)):
+            wanters.set(acceptor, [].asMap().diverge())
+        def wants := wanters.fetch(acceptor)
+        wants.set(giftName, promres)
+        return promise
```
Comment (MostAwesomeDude) [inline airbrus.mt]: Since all branches of control flow return `promise`, consider refactoring to have fewer returns.

File: airbrus.mt [hunk 18]
```diff
@@ -68,6 +68,55 @@ def UTF8JSON :DeepFrozen := composeCodec(UTF8, JSON)
 def makeLineTube() as DeepFrozen:
     return makePumpTube(makeSplitPump(b`$\n`))
 
+def make_giftExchange() :Any as DeepFrozen:
+    def givers  := [].asMap().diverge()
+    def wanters := [].asMap().diverge()
+        
+    def give(giver :Str, acceptor :Str, giftName :Str, gift :Any) :Void :
```
Comment (MostAwesomeDude) [inline airbrus.mt]: Shouldn't this function and the following `accept` function be rolled into the methods which call them further down in the code?

File: airbrus.mt (hunk 7 above)
Comment (MostAwesomeDude) [reply airbrus.mt]: This is a bug; `return` is not legal inside `fn` on its own, and the code will end up calling the (stale) `__return` ejector from `main`'s frame.

**Review by MostAwesomeDude — CHANGES_REQUESTED** [review]
I found some issues.

File: airbrus.mt (hunk 8 above)
Comment (zarutian) [reply airbrus.mt]: snit fixed.

File: airbrus.mt (hunk 18 above)
Comment (zarutian) [reply airbrus.mt]: Rolled them in. Started out this way due to habit of writing modules in Lua in this style.

File: airbrus.mt (hunk 10 above)
Comment (zarutian) [reply airbrus.mt]: closing parenthesis added.

File: airbrus.mt (hunk 11 above)
Comment (zarutian) [reply airbrus.mt]: Changed giftBox to be the result of _makeFinalSlot(). This is needed for when gifts are themselfs promises.

File: airbrus.mt (hunk 14 above)
Comment (zarutian) [reply airbrus.mt]: Implict returns of control flow rub me the wrong way. Harmless to leave in.

File: airbrus.mt (hunk 12 above)
Comment (zarutian) [reply airbrus.mt]: Yes, fixed.

File: airbrus.mt (hunk 15 above)
Comment (zarutian) [reply airbrus.mt]: yet another closing parenthesis missing. Fixed.

**Comment by zarutian** [comment]
I request a review of changes done so far.

**Pull request merged**